### PR TITLE
Fix path resolver when path start with directory separator

### LINKED ISSUE #508: Directory file paths have double DIRECTORY_SEPARATOR
| Q                | A
| ---------------- | -----
| Bug report?      | yes
| Feature request? | no
| Library version  | 2.0.0

When `/app` is set as the execution directory, `$this->directoryList` in `NunoMaduro\PhpInsights\Infrastructure\Repositories\LocalFilesRepository::within()` keeps a double path of `//app` and DIRECTORY_SEPARATOR `//app` and DIRECTORY_SEPARATOR will hold a double path and will not match the runtime file path for `NunoMaduro\PhpInsights\Domain\Insights\Insight::shouldSkipFile()` etc.

---
``` php
namespace NunoMaduro\PhpInsights\Domain\Insights;

final class InsightCollectionFactory
{
    public function get(array $metrics, OutputInterface $consoleOutput): InsightCollection
    {
        $paths = $this->config->getPaths(); // => /app
        $commonPath = $this->config->getCommonPath();
```

``` php
namespace NunoMaduro\PhpInsights\Infrastructure\Repositories;

final class LocalFilesRepository implements FilesRepository
{
    public function within(array $paths, array $exclude = []): FilesRepository
    {
        foreach ($paths as $path) {
            $pathInfo = pathinfo($path);

            if (! is_dir($path) && is_file($path)) {
                $this->fileList['dirname'][] = $pathInfo['dirname'];
                $this->fileList['basename'][] = $pathInfo['basename'];
                $this->fileList['full_path'][] = $pathInfo['dirname'] . DIRECTORY_SEPARATOR . $pathInfo['basename'];
            } else {
               $this->directoryList[] = $pathInfo['dirname'] . DIRECTORY_SEPARATOR . $pathInfo['basename']; // => //app
            }
        }
```

``` php
namespace NunoMaduro\PhpInsights\Domain\Insights;

abstract class Insight implements InsightContract
{
    final protected function shouldSkipFile(string $file): bool
    {
        var_dump([
            $file,
            array_keys($this->excludedFiles),
        ]);

        return \array_key_exists($file, $this->excludedFiles);
        // $file => //app/src/file.php
        // $this->excludedFIles => { '/app/src/file.php' => .. }
    }
```

``` php
// phpinsight.php [config]
return [
    'config' => [
        XXX::class => [
            'exclude' => [
                'src/file.php',
            ],
```

### PR DESCRIPTION
**Pull request by 50bhan**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| Fixed tickets | #508

This PR will close #508.

### CODE REVIEW

File: tests/Application/PathResolverTest.php [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+        $paths = PathResolver::resolve($input);
+
+        self::assertEquals($paths, [$targetPath]);
+    }
+
+    public function testResolvePathWithDirectorySeparator(): void
+    {
+        $path = '/app';
+        $sanitizedPath = getcwd() . $path;
```
Comment (Jibbarth) [inline tests/Application/PathResolverTest.php:37]: I think it's incorrect. 

If I provide an absolute path (like `/app`), this shouldn't be resolved as `current working dir + app`.

**Review by Jibbarth — CHANGES_REQUESTED** [review]
Hey @50bhan, just tested your implementation. 

Given PHPInsights is installed globally, I execute it from my HOME directory.

```
$ phpinsights analyse /tmp/app 
```
This work and launch the analyse on the expected directory

```
$ phpinsights analyse /tmp
```
This doesn't work, raise the following exception : 
```
  The "/home/barth/tmp" directory does not exist.  
```

So we have a regression here

**Comment by Jibbarth** [comment]
@50bhan, for me, the bug is here https://github.com/nunomaduro/phpinsights/blob/master/src/Infrastructure/Repositories/LocalFilesRepository.php#L66

Could be resolved like this 

```php
$this->directoryList[] = rtrim($pathInfo['dirname'], DIRECTORY_SEPARATOR) . DIRECTORY_SEPARATOR . $pathInfo['basename'];
```

But I'm not sure it'll work on Windows.

**Comment by 50bhan** [comment]
> @50bhan, for me, the bug is here https://github.com/nunomaduro/phpinsights/blob/master/src/Infrastructure/Repositories/LocalFilesRepository.php#L66
> 
> Could be resolved like this
> 
> ```
> $this->directoryList[] = rtrim($pathInfo['dirname'], DIRECTORY_SEPARATOR) . DIRECTORY_SEPARATOR . $pathInfo['basename'];
> ```
> 
> But I'm not sure it'll work on Windows.

Actually, this isn't the correct workaround. At first, I did the exact change that you've made. But the problem still remains:

Screen Shot 2021-07-04 at 23 47 47

**Comment by Jibbarth** [comment]
@50bhan i think it's the expected behaviour. When we pass an absolute path, it should be used. 

But in the LocalFileRepository, it transform `/app` into `//app`

**Comment by 50bhan** [comment]
@Jibbarth I was confused about it. If this is the case, the fix is easy as you said.

**Comment by Jibbarth** [comment]
Great ! If I found a Windows, I'll test against

**Comment by 50bhan** [comment]
@Jibbarth Thanks, it would be awesome. I updated the PR.

**Review by Jibbarth — APPROVED** [review]
Tested on Windows, bug was also present and now it's fixed :tada: 

Thank you @50bhan :pray:

**Pull request merged**